File: 08.BasicDataStructure/Sample Code/BasicDataStructureObjc/BasicDataStructureObjc/ViewController.m
#import "ViewController.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ViewController

#pragma mark - UITableViewDataSource Methods

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 50;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *identifier = @"PlainCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    
    if ([indexPath row] == 0)
    {
        [[cell textLabel] setText:@"Red"];
    }
    else if ([indexPath row] == 1)
    {
        [[cell textLabel] setText:@"Orange"];
    }
    else if ([indexPath row] == 2)
    {
        [[cell textLabel] setText:@"Yellow"];
    }
    else if ([indexPath row] == 3)
    {
        [[cell textLabel] setText:@"Green"];
    }
    else if ([indexPath row] == 4)
    {
        [[cell textLabel] setText:@"Blue"];
    }
    else if ([indexPath row] == 5)
    {
        [[cell textLabel] setText:@"Magenta"];
    }

    return cell;
}

#pragma mark - UITableViewDelegate Methods

- (void)tableView:(UITableView *)tableView
didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if ([indexPath row] == 0)
    {
        NSLog(@"Red");
    }
    else if ([indexPath row] == 1)
    {
        NSLog(@"Orange");
    }
    else if ([indexPath row] == 2)
    {
        NSLog(@"Yellow");
    }
    else if ([indexPath row] == 3)
    {
        NSLog(@"Green");
    }
    else if ([indexPath row] == 4)
    {
        NSLog(@"Blue");
    }
    else if ([indexPath row] == 5)
    {
        NSLog(@"Magenta");
    }
}

@end

File: 14.MultipleRowSelection/Sample Code/MultipleCheckmarksSwift/MultipleCheckmarksSwift/ViewController.swift
import UIKit

// http://stackoverflow.com/a/30724543
extension Array where Element : Equatable {
    mutating func removeObject(object : Generator.Element) {
        if let index = self.indexOf(object) {
            self.removeAtIndex(index)
        }
    }
}

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate
{
    @IBOutlet weak var tableView: UITableView!
    var contents = [String]()
    var selectedIndexPaths = [NSIndexPath]()
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        contents = [ "Red", "Orange", "Yellow", "Green", "Blue", "Magenta" ]
    }
    
    @IBAction func reloadButtonTapped(sender: UIButton)
    {
        tableView.reloadData()
    }
    
    func populateCell(cell: UITableViewCell, indexPath: NSIndexPath)
    {
        let rowValue = contents[indexPath.row];
        
        cell.textLabel?.text = rowValue
        
        if (selectedIndexPaths.contains(indexPath))
        {
            cell.accessoryType = .Checkmark
        }
        else
        {
            cell.accessoryType = .None
        }
    }
    
    // MARK: - UITableViewDataSource Methods
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return contents.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        let identifier = "PlainCell"
        
        var optionalCell = tableView.dequeueReusableCellWithIdentifier(identifier)
        if (optionalCell == nil)
        {
            optionalCell = UITableViewCell.init(style: .Default, reuseIdentifier: identifier)
        }
        
        let cell = optionalCell!
        
        populateCell(cell, indexPath: indexPath)
        
        return cell
    }
    
    // MARK: - UITableViewDelegate Method
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
    {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        if (selectedIndexPaths.contains(indexPath))
        {
            selectedIndexPaths.removeObject(indexPath)
        }
        else
        {
            selectedIndexPaths.append(indexPath)
        }
        
        let cell = tableView.cellForRowAtIndexPath(indexPath)!
        populateCell(cell, indexPath: indexPath)
    }
}